# import modules

In [137]:
from preparation import get_train_valid

import pandas as pd

# Define constants 

In [138]:
TARGET = "Transported"

# Load data

In [139]:
train_data, valid_data = get_train_valid()
target = train_data[TARGET]
train_data = train_data.drop(columns=TARGET)

In [140]:
display(train_data.head(3))
display(valid_data.head(3))

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck             Name  
0          0.0        0.0           0.0     0.0     0.0  Maham Ofracculy  
1        109.0        9.0          25.0   549.0    44.0     Juanna Vines  
2         43.0     3576.0           0.0  6715.0    49.0    Altark Susent

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0013_01      Earth      True  G/3/S  TRAPPIST-1e  27.0  False   
1     0018_01      Earth     False  F/4/S  TRAPPIST-1e  19.0  False   
2     0019_01     Europa      True  C/0/S  55 Cancri e  31.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck             Name  
0          0.0        0.0           0.0     0.0     0.0  Nelly Carsoning  
1          0.0        9.0           0.0  2823.0     0.0   Lerome Peckers  
2          0.0        0.0           0.0     0.0     0.0  Sabih Unhearfus

In [141]:
display(train_data.describe())
display(valid_data.describe())

Age   RoomService     FoodCourt  ShoppingMall           Spa  \
count  8514.000000   8512.000000   8510.000000   8485.000000   8510.000000   
mean     28.827930    224.687617    458.077203    173.729169    311.138778   
std      14.489021    666.717663   1611.489240    604.696458   1136.705535   
min       0.000000      0.000000      0.000000      0.000000      0.000000   
25%      19.000000      0.000000      0.000000      0.000000      0.000000   
50%      27.000000      0.000000      0.000000      0.000000      0.000000   
75%      38.000000     47.000000     76.000000     27.000000     59.000000   
max      79.000000  14327.000000  29813.000000  23492.000000  22408.000000   

             VRDeck  
count   8505.000000  
mean     304.854791  
std     1145.717189  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%       46.000000  
max    24133.000000

Age   RoomService     FoodCourt  ShoppingMall           Spa  \
count  4186.000000   4195.000000   4171.000000   4179.000000   4176.000000   
mean     28.658146    219.266269    439.484296    177.295525    303.052443   
std      14.179072    607.011289   1527.663045    560.821123   1117.186015   
min       0.000000      0.000000      0.000000      0.000000      0.000000   
25%      19.000000      0.000000      0.000000      0.000000      0.000000   
50%      26.000000      0.000000      0.000000      0.000000      0.000000   
75%      37.000000     53.000000     78.000000     33.000000     50.000000   
max      79.000000  11567.000000  25273.000000   8292.000000  19844.000000   

             VRDeck  
count   4197.000000  
mean     310.710031  
std     1246.994742  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%       36.000000  
max    22272.000000

In [142]:
display(train_data.describe(include="O"))
display(valid_data.describe(include="O"))

PassengerId HomePlanet CryoSleep    Cabin  Destination    VIP  \
count         8693       8492      8476     8494         8511   8490   
unique        8693          3         2     6560            3      2   
top        0001_01      Earth     False  G/734/S  TRAPPIST-1e  False   
freq             1       4602      5439        8         5915   8291   

                  Name  
count             8493  
unique            8473  
top     Gollux Reedall  
freq                 2

PassengerId HomePlanet CryoSleep    Cabin  Destination    VIP  \
count         4277       4190      4184     4177         4185   4184   
unique        4277          3         2     3265            3      2   
top        0013_01      Earth     False  G/160/P  TRAPPIST-1e  False   
freq             1       2263      2640        8         2956   4110   

              Name  
count         4183  
unique        4176  
top     Cints Erle  
freq             2

# Fill none values

In [143]:
train_data.isna().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
dtype: int64

Home planet is unknown in 201 rows. Let's consider "PassengerId". If passengers in one group and some of them have None value, we'll fill this value the same, as known passanger

Create "group" column that include "gggg" in "gggg_pp" form "PassengerId"

In [144]:
train_data["group"] = train_data["PassengerId"].map(lambda x: x[:4]).astype(int)
train_data["members_in_group"] = train_data["PassengerId"].map(lambda x: x[5:]).astype(int)
train_data.head(3)

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck             Name  \
0          0.0        0.0           0.0     0.0     0.0  Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0     Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0    Altark Susent   

   group  members_in_group  
0      1                 1  
1      2                 1  
2      3                 1

Let's check unique groups. 6217, so 8693 - 6217 = 2476 repeated groups in data

In [145]:
train_data[["group", "members_in_group"]].nunique()

group               6217
members_in_group       8
dtype: int64

Fill NAN values

In [148]:
def filler(x):
    if x.isna().any():
        value = "Earth"
    else:
        value = x.value_counts().keys()[0]
    return x.fillna(value)

train_data["HomePlanet"] = train_data.groupby("group")["HomePlanet"].transform(filler)

In [150]:
train_data.isna().sum()

PassengerId           0
HomePlanet            0
CryoSleep           217
Cabin               199
Destination         182
Age                 179
VIP                 203
RoomService         181
FoodCourt           183
ShoppingMall        208
Spa                 183
VRDeck              188
Name                200
group                 0
members_in_group      0
dtype: int64

In [29]:
# num_columns = train_data.dtypes[train_data.dtypes == float].keys().tolist()
# cat_columns = ["HomePlanet", "CryoSleep", "Destination", "VIP"]
# obj_columns = ["PassengerId", "Cabin", "Name"]

In [32]:
# (len(num_columns) + len(cat_columns) + len(obj_columns)) == len(train_data.keys())

True

In [34]:
# train_data[cat_columns].nunique()

HomePlanet     3
CryoSleep      2
Destination    3
VIP            2
dtype: int64